### PR TITLE
Update ConnectorMaterializedViewDefinition

### PR DESCRIPTION
**Pull request by raunaqmorarka**
Decoupled ConnectorMaterializedViewDefinition from the way MV is stored by iceberg connector.
This allows evolving ConnectorMaterializedViewDefinition class without affecting backward compatibility with existing MV stored in iceberg.

Allow storage table to be in a different catalog and schema from the one used for storing the
materialized view definition.

### CODE REVIEW

**Review by anjalinorwood — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -858,7 +863,8 @@ public void createMaterializedView(ConnectorSession session, SchemaTableName vie
                 .setParameters(viewProperties)
                 .withStorage(storage -> storage.setStorageFormat(VIEW_STORAGE_FORMAT))
                 .withStorage(storage -> storage.setLocation(""))
-                .setViewOriginalText(Optional.of(encodeMaterializedViewData(definition)))
+                .setViewOriginalText(Optional.of(encodeMaterializedViewData(
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:866]: I think we should extract internal and backward compatible representation of MV for iceberg and use it for serde. This way we won't break any backward compatibility.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 2]
```diff
@@ -1060,8 +1069,11 @@ else if (strings.size() != 2) {
             return viewToken;
         }
 
-        String storageTableName = Optional.ofNullable(materializedViewDefinition.get().getStorageTable()).orElse("");
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1063]: was it a bug?

Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1063]: no, a change is needed here due to change in storageTable field in ConnectorMaterializedViewDefinition

Comment (sopel39) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1063]: `.orElse("");` seems problematic since `getIcebergTable` was called next

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java [hunk 3]
```diff
@@ -49,7 +49,7 @@ public ConnectorMaterializedViewDefinition(
         this.schema = requireNonNull(schema, "schema is null");
         this.columns = List.copyOf(requireNonNull(columns, "columns is null"));
         this.comment = requireNonNull(comment, "comment is null");
-        this.owner = Optional.of(requireNonNull(owner, "owner is null"));
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java]: this really belongs to other commit

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java [hunk 4]
```diff
@@ -13,8 +13,6 @@
  */
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java:13]: Extract this into three commits:
1. Commit that decouples SPI MV from Iceberg MV
2. Commit that simplifies SPI MV (including non-optional owner)
3. Commit that removes Json annotations in SPI MV. This could be separate PR potentially to let other maintainers to review as it changes how we should model other POJO SPI classes.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/*
+ * Serializable version of ConnectorMaterializedViewDefinition stored by iceberg connector
+ */
+public class IcebergMaterializedViewDefinition
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java:34]: keep the abstraction simple. If only iceberg is using this and it had broken serde, move it to `trino-iceberg`. If we think we might use it in Hive, move it to `trino-hive` and call it `TrinoMaterializedViewDefinition`.

But I don't think we will support Trino MVs in `trino-hive` soon since `storageTable` is kind of implementation specific

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/TrinoMaterializedViewSerde.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import io.airlift.json.JsonCodec;
+
+import java.util.Base64;
+
+import static io.trino.plugin.hive.HiveErrorCode.HIVE_INVALID_VIEW_DATA;
+import static io.trino.plugin.hive.util.HiveUtil.checkCondition;
+import static java.util.Objects.requireNonNull;
+
+public class TrinoMaterializedViewSerde<T>
```
Comment (sopel39) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/TrinoMaterializedViewSerde.java:11]: `T` could be anything, not only MV

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+    @JsonCreator
+    public IcebergMaterializedViewDefinition(
+            @JsonProperty("originalSql") String originalSql,
+            @JsonProperty("storageTable") @Nullable String storageTable,
+            @JsonProperty("catalog") Optional<String> catalog,
+            @JsonProperty("schema") Optional<String> schema,
+            @JsonProperty("columns") List<Column> columns,
+            @JsonProperty("comment") Optional<String> comment,
+            @JsonProperty("owner") Optional<String> owner,
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java:53]: I think this one can be `String` and still work with "old" json-s. (for the sake of order/cleanness)

**Comment by sopel39** [comment]
small comments

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java [hunk 8]
```diff
@@ -46,33 +47,14 @@ public ConnectorMaterializedViewDefinition(
             @JsonProperty("comment") Optional<String> comment,
             @JsonProperty("owner") String owner,
             @JsonProperty("properties") Map<String, Object> properties)
-    {
-        this(originalSql, requireNonNull(storageTable, "storageTable is null").orElse(null), catalog, schema, columns, comment, Optional.of(owner), properties);
-    }
-
-    /*
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java:53]: Please make `ConnectorMaterializedViewDefinition` cleanup separate from decoupling in Iceberg

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java [hunk 9]
```diff
@@ -37,6 +37,7 @@
     private final Optional<String> owner;
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java]: after decoupling `owner` doesn't have to be `Optional`

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java [hunk 10]
```diff
@@ -46,33 +47,14 @@ public ConnectorMaterializedViewDefinition(
...
             @JsonProperty("properties") Map<String, Object> properties)
-    {
-        this(originalSql, requireNonNull(storageTable, "storageTable is null").orElse(null), catalog, schema, columns, comment, Optional.of(owner), properties);
-    }
-
-    /*
-     * This constructor is for JSON deserialization only. Do not use.
-     */
-    // TODO: Simplify this constructor and getters: https://github.com/trinodb/trino/issues/7537
-    @Deprecated
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java:57]: these changes belong to `Update ConnectorMaterializedViewDefinition`

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java [hunk 11]
```diff
@@ -35,16 +33,15 @@
     private final String owner;
     private final Map<String, Object> properties;
 
-    @JsonCreator
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java]: Let's have this commit as a follow up for other maintainers to review.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+import static io.trino.plugin.hive.HiveErrorCode.HIVE_INVALID_VIEW_DATA;
+import static io.trino.plugin.hive.util.HiveUtil.checkCondition;
+import static java.util.Objects.requireNonNull;
+
+/*
+ * Serializable version of ConnectorMaterializedViewDefinition stored by iceberg connector
+ */
+public class IcebergMaterializedViewDefinition
+{
+    private static final String MATERIALIZED_VIEW_PREFIX = "/* Presto Materialized View: ";
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java:42]: Changing it now would require changes to `decodeMaterializedViewData` to accept both old and new prefix.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+    private static final JsonCodec<IcebergMaterializedViewDefinition> materializedViewCodec =
+            new JsonCodecFactory(new ObjectMapperProvider()).jsonCodec(IcebergMaterializedViewDefinition.class);
+
+    private final String originalSql;
+    private final String storageTable;
+    private final Optional<String> catalog;
+    private final Optional<String> schema;
+    private final List<Column> columns;
+    private final Optional<String> comment;
+    private final String owner;
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java:54]: Yes, that works

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 14]
```diff
@@ -861,7 +866,8 @@ public void createMaterializedView(ConnectorSession session, SchemaTableName vie
                 .withStorage(storage -> storage.setStorageFormat(VIEW_STORAGE_FORMAT))
                 .withStorage(storage -> storage.setLocation(""))
                 .setViewOriginalText(Optional.of(
-                        encodeMaterializedViewData(fromConnectorMaterializedViewDefinition(definition))))
+                        encodeMaterializedViewData(
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:869]: MV was encoded created previously without `storage table`. Properties are also not really used (we derive them from table). I think we could remove `IcebergMaterializedViewDefinition` from `storageTable` and `properties` from `IcebergMaterializedViewDefinition`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 15]
```diff
@@ -992,15 +998,19 @@ private boolean isMaterializedView(Table table)
         }
 
         Table materializedView = tableOptional.get();
+        Optional<String> storageTable = Optional.ofNullable(materializedView.getParameters().get(STORAGE_TABLE));
+        if (storageTable.isEmpty()) {
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1002]: I don't think it should ever by empty, `createMaterializedView` always adds storage table.
`checkState` instead?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 16]
```diff
@@ -1065,8 +1075,11 @@ else if (strings.size() != 2) {
             return viewToken;
         }
 
-        String storageTableName = Optional.ofNullable(materializedViewDefinition.get().getStorageTable()).orElse("");
-        org.apache.iceberg.Table icebergTable = getIcebergTable(metastore, hdfsEnvironment, session, new SchemaTableName(name.getSchemaName(), storageTableName));
+        Optional<SchemaTableName> storageTableName = materializedViewDefinition.get().getStorageTable().map(CatalogSchemaTableName::getSchemaTableName);
+        if (storageTableName.isEmpty()) {
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1079]: I don't think `storageTable` can be empty, `checkState` instead?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 14 above)
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:869]: We keep storage table in Hive view properties (it was not encoded previously as part of `ConnectorMaterializedViewDefinition`). We also don't use MV properties, but rather table properties. Maybe let's encode just needed fields in `IcebergMaterializedViewDefinition`?

Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java]: I've removed storageTable from IcebergMaterializedViewDefinition now

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+        if (catalog.isEmpty() && schema.isPresent()) {
+            throw new IllegalArgumentException("catalog must be present if schema is present");
+        }
+        if (columns.isEmpty()) {
+            throw new IllegalArgumentException("columns list is empty");
+        }
+    }
+
+    public static IcebergMaterializedViewDefinition fromConnectorMaterializedViewDefinition(ConnectorMaterializedViewDefinition definition)
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java:76]: move `fromConnectorMaterializedViewDefinition` above `IcebergMaterializedViewDefinition` constructor

**Review by sopel39 — APPROVED** [review]

**Comment by sopel39** [comment]
small comment

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+        }
+
+        @Override
+        public String toString()
+        {
+            return name + " " + type;
+        }
+    }
+
+    public static String encodeMaterializedViewData(IcebergMaterializedViewDefinition definition)
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewDefinition.java:171]: move this methods to the top of class along with `fromConnectorMaterializedViewDefinition` method
